File: rust/src/_3_longest_substring_without_repeating_characters.rs
/*
 * @lc app=leetcode id=3 lang=rust
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
use std::cmp::max;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let n_chars = s.len();
        if n_chars < 2 {
            return n_chars as i32;
        }
        let s: Vec<char> = s.chars().collect();
        let mut l = 0;
        let mut r = 0;
        let mut max_len = 0;
        while r < n_chars {
            if let Some(i) = s[l..r].iter().position(|&c| c == s[r]) {
                max_len = max(max_len, r - l);
                l += i + 1;
                r += 1;
            } else {
                r += 1;
            }
        }
        max_len = max(max_len, r - l);
        max_len as i32
    }
}
// @lc code=end

struct Solution;

File: typescript/src/_3_longest_substring_without_repeating_characters.ts
/*
 * @lc app=leetcode id=3 lang=typescript
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
function lengthOfLongestSubstring(s: string): number {
    const nChars = s.length;
    if (nChars < 2) {
        return nChars;
    }
    const arr: string[] = s.split('');
    let l = 0;
    let r = 0;
    let maxLen = 0;
    while (r < nChars) {
        const i = arr.slice(l, r).indexOf(arr[r]!);
        if (-1 < i) {
            maxLen = Math.max(maxLen, r - l);
            l += i + 1;
            r += 1;
        } else {
            r += 1;
        }
    }
    maxLen = Math.max(maxLen, r - l);
    return maxLen
};
// @lc code=end
